Fix error caused by weakly referenced package not specifying version …

loonghao opened the pull request:
This PR is try to fix #1711

JeanChristopheMorinPerso commented:
@loonghao Thanks for creating this PR. Do you think you could add a test for these lines: https://github.com/AcademySoftwareFoundation/rez/pull/1712/files#diff-1ad6984e6ce49221a86620d55092bacae0b32b2d6c9401295c655c9d3e72a1a0R443-R446?

loonghao commented:
> @loonghao Thanks for creating this PR. Do you think you could add a test for these lines: https://github.com/AcademySoftwareFoundation/rez/pull/1712/files#diff-1ad6984e6ce49221a86620d55092bacae0b32b2d6c9401295c655c9d3e72a1a0R443-R446?

@JeanChristopheMorinPerso 
I adjusted some tests and found that I still couldn't cover these lines of code.
I think since they are all expected they won't raise a TypeError, I removed the update on these lines.

isohedronpipeline commented on src/rez/tests/test_solver.py:
```diff
@@ -256,6 +256,21 @@ def test_12_missing_variant_requires(self):
+                     'test_variant_split_end-3.0[0]',
+                     'test_variant_split_mid2-2.0[1]'])
+
+    def test_13_resolve_weakly_reference_variant(self):
```
Should this be `test_14_...`?

isohedronpipeline commented on src/rez/package_order.py:
```diff
@@ -124,23 +124,28 @@ def sort_key(self, package_name, version_like):
+            # Thus, to maintain the original order,
+            # we return the same object for all None values.
+            return 0
         if isinstance(version_like, VersionRange):
             return tuple(self.sort_key(package_name, bound) for bound in version_like.bounds)
-        elif isinstance(version_like, _Bound):
+        if isinstance(version_like, _Bound):
             return (self.sort_key(package_name, version_like.lower),
                     self.sort_key(package_name, version_like.upper))
-        elif isinstance(version_like, _LowerBound):
+        if isinstance(version_like, _LowerBound):
             inclusion_key = -2 if version_like.inclusive else -1
             return self.sort_key(package_name, version_like.version), inclusion_key
-        elif isinstance(version_like, _UpperBound):
+        if isinstance(version_like, _UpperBound):
             inclusion_key = 2 if version_like.inclusive else 1
             return self.sort_key(package_name, version_like.version), inclusion_key
-        elif isinstance(version_like, Version):
+        if isinstance(version_like, Version):
             # finally, the bit that we actually use the sort_key_implementation for.
             return FallbackComparable(
                 self.sort_key_implementation(package_name, version_like), version_like)
-        else:
-            raise TypeError(version_like)
+        raise TypeError(version_like)
```
We want explicitly `if version_like is None:` since we want it to fail if some other non Version type is passed in whose `__bool__` returns `False`.

Since a `None` case is less likely than a version range, this does belong at the bottom.

I'm also not sure if we need to refactor the rest of it...
Also, I'm not sure why we're removing the TypeError.  Then it'll just return `None` which is not what we want either.

loonghao replied:
Change the elif statement to an if statement. Because each if and elif statement block has a return statement, there will be no situation where multiple conditional blocks are executed.

I see the test is failing and I now revert the `TypeError`

loonghao replied:
> Since a None case is less likely than a version range, this does belong at the bottom.
I will update now to keep it as is

JeanChristopheMorinPerso commented on src/rez/package_order.py:
```diff
@@ -126,21 +126,26 @@ def sort_key(self, package_name, version_like):
                 self.sort_key_implementation(package_name, version_like), version_like)
-        else:
-            raise TypeError(version_like)
+        if version_like is None:
```
@loonghao Can you adjust the type in the docstring to mention that None is a valid type please?

JeanChristopheMorinPerso commented on src/rez/tests/test_solver.py:
```diff
@@ -256,6 +256,21 @@ def test_12_missing_variant_requires(self):
         config.override("error_on_missing_variant_requires", False)
         self._solve(["missing_variant_requires"], ["nada[]", "missing_variant_requires-1[1]"])
 
+    def test_13_resolve_weakly_reference_requires(self):
```
Can we add a comment to say something more clearly describe what's being tested please?

JeanChristopheMorinPerso commented on src/rez/tests/test_solver.py:
```diff
@@ -256,6 +256,21 @@ def test_12_missing_variant_requires(self):
+                     'test_variant_split_end-3.0[0]',
+                     'test_variant_split_mid2-2.0[1]'])
+
+    def test_14_resolve_weakly_reference_variant(self):
```
Same here.

JeanChristopheMorinPerso reviewed (approved):
Thank you @loonghao and @isohedronpipeline!
